### DISCUSSION TITLE
TrenchBroom map making

### DISCUSSION THREAD

**Discussion by ImTander** [nzp-team/nzportable#675] [General]
hello all, I've been spending about a week trying to teach myself to make nzp maps. I started with worldcraft bc that al the tutorial I could find. Just recently started with trench broom but i cant seem to actually play or compile/test my maps and I'm honestly about ready to give up. I keep getting linked to the trenchbroom tut series which I'm watching but he's making maps for quake, not nzp. So I cant find any clear steps an how to compile an run a map with trench broom. I've looked up multiple tutorials to try to get running and every time I compile I'm hit with "cannot convert value 'undefined' of type 'undefined' to type 'string'. Like I said, I've looked around everywhere and I'm not getting any help, everything is 5+ years old. PLZ HELP

**Comment by MotoLegacy** [comment]
The compilation steps are exactly the same between Quake and NZ:P, with the caveat that you use the VHLT (https://docs.nzp.gay/mapping/trenchbroom-setup) compiler to run instead of ericw-tools. TrenchBroom's manual has a more in-depth description of how the compilation pipeline works and gives an example: https://trenchbroom.github.io/manual/latest/#compiling_maps

The NZ:P Docs are actively maintained, and link to both of these pages:
https://docs.nzp.gay/

**Reply by ImTander** [reply]
![Desktop Screenshot 2024 03 03 - 21 05 12 31](https://github.com/nzp-team/nzportable/assets/161973907/d6662e6a-2898-4b88-a5a7-0bc1fb640103)
I believe I'm using the half life tools, no? I tried following the video exactly but I keep hitting that error

**Reply by lavenderdotpet** [reply]
could u send a screenshot of ur compile menu in trenchbroom

**Reply by ImTander** [reply]
I downloaded the tools cypress released last night and was still trying to figure out what was wrong so I might've messed something up accidentally. Here's how everything looks now 
![Desktop Screenshot 2024 03 05 - 12 24 19 95](https://github.com/nzp-team/nzportable/assets/161973907/43045121-5071-48c3-a149-fde273d28c34)
![Desktop Screenshot 2024 03 05 - 12 24 55 47](https://github.com/nzp-team/nzportable/assets/161973907/6f8d4cc4-074d-4029-93ec-95f4e313a242)
![Desktop Screenshot 2024 03 05 - 12 25 03 57](https://github.com/nzp-team/nzportable/assets/161973907/cbe42d0c-97e0-4b17-9acd-470618d6756e)

Thanks for the help!

**Reply by lavenderdotpet** [reply]
whats the output when u press compile

**Comment by ImTander** [comment]
#### Using working directory 'C:\Users\ImTan\OneDrive\Desktop\NZP\nzp\maps'
#### Exporting map file 'C:\Users\ImTan\OneDrive\Desktop\NZP\nzp\maps\trenchbroom-learning-compile.map'
#### Executing '"C:\TrenchBroom\tools\hlcsg.exe" C:\Users\ImTan\OneDrive\Desktop\NZP\nzp\maps/trenchbroom-learning.map -nowadtextures'
hlcsg v3.4 VL34 64-bit (Mar  4 2024)
Zoner's Half-Life Compilation Tools -- Custom Build
Based on code modifications by Sean 'Zoner' Cavanaugh
Based on Valve's version, modified with permission.
Submit detailed bug reports to (vluzacn@163.com)
-----  BEGIN  hlcsg -----
Command line: C:\TrenchBroom\tools\hlcsg.exe C:\Users\ImTan\OneDrive\Desktop\NZP\nzp\maps/trenchbroom-learning.map -nowadtextures 
Arguments: C:\Users\ImTan\OneDrive\Desktop\NZP\nzp\maps/trenchbroom-learning.map -nowadtextures 
Entering C:\Users\ImTan\OneDrive\Desktop\NZP\nzp\maps\trenchbroom-learning.map

Current hlcsg Settings
Name                 |  Setting  |  Default
---------------------|-----------|-------------------------
threads               [      16 ] [  Varies ]
verbose               [     off ] [     off ]
log                   [      on ] [      on ]
reset logfile         [      on ] [      on ]
developer             [       0 ] [       0 ]
chart                 [     off ] [     off ]
estimate              [     off ] [     off ]
max texture memory    [ 33554432 ] [ 33554432 ]
max lighting memory   [ 50331648 ] [ 50331648 ]
priority              [  Normal ] [  Normal ]

noclip                [     off ] [     off ]
null texture stripping[      on ] [      on ]
clipnode economy mode [     off ] [     off ]
clip hull type        [  simple ] [  simple ]
onlyents              [     off ] [     off ]
wadtextures           [     off ] [      on ]
skyclip               [      on ] [      on ]
hullfile              [    None ] [    None ]
wad configuration file[    None ] [    None ]
wad.cfg group name    [    None ] [    None ]
nullfile              [    None ] [    None ]
nullify trigger       [      on ] [      on ]
min surface area      [   0.000 ] [   0.000 ]
brush union threshold [   0.000 ] [   0.000 ]
map scaling           [    None ] [    None ]
light name optimize   [      on ] [      on ]
convert game_text     [      on ] [      on ]

Using mapfile wad configuration
Wadinclude list :
[zhlt.wad]

CreateBrush:
 (0.00 seconds)
CSGBrush:
 (0.01 seconds)

Including Wadfile: C:/Users/ImTan/OneDrive/Desktop/NZP/map creation/assets-newest/source/textures/wad/4all.wad
 - Contains 0 used textures, 0.00 percent of map (28 textures in wad)
Including Wadfile: C:/Users/ImTan/OneDrive/Desktop/NZP/map creation/assets-newest/source/textures/wad/b1oodv4.wad
 - Contains 0 used textures, 0.00 percent of map (14 textures in wad)
Including Wadfile: C:/Users/ImTan/OneDrive/Desktop/NZP/map creation/assets-newest/source/textures/wad/chalk_drawings.wad
 - Contains 0 used textures, 0.00 percent of map (14 textures in wad)
Including Wadfile: C:/Users/ImTan/OneDrive/Desktop/NZP/map creation/assets-newest/source/textures/wad/Dung3on.wad
 - Contains 0 used textures, 0.00 percent of map (14 textures in wad)
Including Wadfile: C:/Users/ImTan/OneDrive/Desktop/NZP/map creation/assets-newest/source/textures/wad/Ju[s]tice_null2.wad
 - Contains 2 used textures, 100.00 percent of map (938 textures in wad)
Including Wadfile: C:/Users/ImTan/OneDrive/Desktop/NZP/map creation/assets-newest/source/textures/wad/Loop.wad
 - Contains 0 used textures, 0.00 percent of map (9 textures in wad)
Including Wadfile: C:/Users/ImTan/OneDrive/Desktop/NZP/map creation/assets-newest/source/textures/wad/nzphlwad.wad
 - Contains 0 used textures, 0.00 percent of map (103 textures in wad)
Including Wadfile: C:/Users/ImTan/OneDrive/Desktop/NZP/map creation/assets-newest/source/textures/wad/perks.wad
 - Contains 0 used textures, 0.00 percent of map (9 textures in wad)
Including Wadfile: C:/Users/ImTan/OneDrive/Desktop/NZP/map creation/assets-newest/source/textures/wad/verruckt.wad
 - Contains 0 used textures, 0.00 percent of map (107 textures in wad)
Including Wadfile: C:/Users/ImTan/OneDrive/Desktop/NZP/map creation/assets-newest/source/textures/wad/warehouse.wad
 - Contains 0 used textures, 0.00 percent of map (1 textures in wad)
Including Wadfile: C:/Users/ImTan/OneDrive/Desktop/NZP/map creation/assets-newest/source/textures/wad/zhlt.wad
 - Contains 0 used textures, 0.00 percent of map (18 textures in wad)

Wad files required to run the map: (None)
Texture usage is at 0.03 mb (of 32.00 mb MAX)
0.03 seconds elapsed

-----   END   hlcsg -----



#### Finished with exit status 0

#### Executing '"C:\TrenchBroom\tools\hlbsp.exe" trenchbroom-learning-compile.map trenchbroom-learning.bsp'
Unknown option "trenchbroom-learning.bsp"
hlbsp v3.4 VL34 64-bit (Mar  4 2024)
Zoner's Half-Life Compilation Tools -- Custom Build
Based on code modifications by Sean 'Zoner' Cavanaugh
Based on Valve's version, modified with permission.
Submit detailed bug reports to (vluzacn@163.com)

-= hlbsp Options =-

    -console #     : Set to 0 to turn off the pop-up console (default is 1)
    -lang file     : localization file
    -leakonly      : Run BSP only enough to check for LEAKs
    -subdivide #   : Sets the face subdivide size
    -maxnodesize # : Sets the maximum portal node size

    -notjunc       : Don't break edges on t-junctions     (not for final runs)
    -nobrink       : Don't smooth brinks                  (not for final runs)
    -noclip        : Don't process the clipping hull      (not for final runs)
    -nofill        : Don't fill outside (will mask LEAKs) (not for final runs)
    -noinsidefill  : Don't fill empty spaces
    -noopt         : Don't optimize planes on BSP write   (not for final runs)
    -noclipnodemerge: Don't optimize clipnodes
    -texdata #     : Alter maximum texture memory limit (in kb)
    -lightdata #   : Alter maximum lighting memory limit (in kb)
    -chart         : display bsp statitics
    -low | -high   : run program an altered priority level
    -nolog         : don't generate the compile logfiles
    -threads #     : manually specify the number of threads to run
    -estimate      : display estimated time during compile
    -nonulltex     : Don't strip NULL faces
    -nohull2       : Don't generate hull 2 (the clipping hull for large monsters and pushables)
    -viewportal    : Show portal boundaries in 'mapname_portal.pts' file
    -verbose       : compile with verbose messages
    -noinfo        : Do not show tool configuration information
    -dev #         : compile with developer message

    mapfile        : The mapfile to compile

#### Finished with exit status 1

#### Executing '"C:\TrenchBroom\tools\hlvis.exe" -threads 15trenchbroom-learning.bsp'
No mapfile specified
hlvis v3.4 VL34 64-bit (Mar  4 2024)
Zoner's Half-Life Compilation Tools -- Custom Build
Based on code modifications by Sean 'Zoner' Cavanaugh
Based on Valve's version, modified with permission.
Submit detailed bug reports to (vluzacn@163.com)

-= hlvis Options =-

    -console #      : Set to 0 to turn off the pop-up console (default is 1)
    -lang file      : localization file
    -full           : Full vis
    -fast           : Fast vis

    -texdata #      : Alter maximum texture memory limit (in kb)
    -lightdata #      : Alter maximum lighting memory limit (in kb)
    -chart          : display bsp statitics
    -low | -high    : run program an altered priority level
    -nolog          : don't generate the compile logfiles
    -threads #      : manually specify the number of threads to run
    -estimate       : display estimated time during compile
    -maxdistance #  : Alter the maximum distance for visibility
    -verbose        : compile with verbose messages
    -noinfo         : Do not show tool configuration information
    -dev #          : compile with developer message

    mapfile         : The mapfile to compile

#### Finished with exit status 1

#### Executing '"C:\TrenchBroom\tools\light.exe" -extra C:\Users\ImTan\OneDrive\Desktop\NZP\nzp\maps/trenchbroom-learning.bsp'
----- Light 1.43 ---- Modified by Bengt Jardrup

Extra 2x2 sampling enabled
File: C:\Users\ImTan\OneDrive\Desktop\NZP\nzp\maps/trenchbroom-learning.bsp
************ ERROR ************
C:\Users\ImTan\OneDrive\Desktop\NZP\nzp\maps/trenchbroom-learning.bsp is version 30, not 29

Elapsed time :  0:00
#### Finished with exit status 1


idk if you wanted the whole thing but there it is lmao

**Reply by MotoLegacy** [reply]
Only provide the `.map` parameter for hlbsp, providing a `.bsp` arg breaks it.

For hlvis, you forgot to add a space after the thread count.

**Comment by ImTander** [comment]
I removed the .map from the hlbsp. Now forgive me and my pea brain, but adding a space into the thread count

-threads ${CPU_COUNT - 1}${MAP_BASE_NAME}.bsp

are you ref to add a space after the number 1 in the {CPU_COUNT - 1} , or a different spot?

sorry as I'm sure this seems really simple for everyone else

**Reply by MotoLegacy** [reply]
`-threads ${CPU_COUNT - 1} ${MAP_BASE_NAME}.bsp`
here it is fixed. it adds a space after the cpu count block and the map base name block.

do note i said remove the .bsp line from hlbsp, not the .map line. you provide the `.map` to `hlbsp` and `hlcsg`, and the `.bsp` to `hlvis` and `hlrad`.

**Reply by ImTander** [reply]
thanks for breaking it down and making it simple for me. I also meant to say bsp earlier not map thats my b. Now I believe everything is set correctly. Also am I meant to set up something for ripent? anyway, when I compile now it ends with "Finished with exit status 1" still. These are two errors I dug from the log

Command line: C:\TrenchBroom\tools\hlvis.exe -threads 15 trenchbroom-learning.bsp 

Arguments: -threads 15 trenchbroom-learning.bsp 

Error: Portal file 'trenchbroom-learning.prt' does not exist, cannot vis the map

and the next one is from the hlrad just saying it can't compile?

Command line: C:\TrenchBroom\tools\hlrad.exe -extra C:\Users\ImTan\OneDrive\Desktop\NZP\nzp\maps/trenchbroom-learning.bsp 

Arguments: -extra C:\Users\ImTan\OneDrive\Desktop\NZP\nzp\maps/trenchbroom-learning.bsp 

>> There was a problem compiling the map.

I did try loading the map to see if anything would happen since its in my maps list but the game just crashes.

I just want to make maps 😅 I probably set up something wrong on accident. Thank you for your time. 

![Desktop Screenshot 2024 03 05 - 13 20 42 30](https://github.com/nzp-team/nzportable/assets/161973907/c9110eb9-ff9f-4b14-91b3-a2a93c19d88b)

**Reply by MotoLegacy** [reply]
Remove the first argument from `hlbsp`, i.e., make sure the only parameter is `{MAP_BASE_NAME}.map`. i think that should be all thats left that is wrong.

**Comment by ImTander** [comment]
thank you so much 😭💜 
![Base Profile Screenshot 2024 03 05 - 13 57 11 69](https://github.com/nzp-team/nzportable/assets/161973907/35c99b8e-b995-42a8-8a77-32f992d8c455)
